packages/flutter/lib/src/cupertino/nav_bar.dart

```diff
@@ -908,6 +930,153 @@ class _LargeTitleNavigationBarSliverDelegate
+    required BoxConstraints navBarConstraints,
+    required double maxExtent,
+    required TextDirection textDirection,
+  }) : _navBarConstraints = navBarConstraints,
+      _maxExtent = maxExtent,
+      _textDirection = textDirection;
+
+  BoxConstraints get navBarConstraints => _navBarConstraints;
+  BoxConstraints _navBarConstraints;
+  set navBarConstraints(BoxConstraints value) {
+    if (_navBarConstraints == value) {
+      return;
+    }
+    _navBarConstraints = value;
+
+    markNeedsLayout();
+  }
+
+  double get maxExtent => _maxExtent;
+  double _maxExtent;
+  set maxExtent(double value) {
+    if (_maxExtent == value) {
+      return;
+    }
+    _maxExtent = value;
+
+    markNeedsLayout();
+  }
+
+  TextDirection get textDirection => _textDirection;
+  TextDirection _textDirection;
+  set textDirection(TextDirection value) {
+    if (_textDirection == value) {
+      return;
+    }
+    _textDirection = value;
+
+    markNeedsLayout();
+  }
+
+  Matrix4? _transform;
+
+  Matrix4? get _effectiveTransform {
+    // Maximum scale lets us prevent the large title
+    // from getting clipped when it's width is greater than
+    // the navigation bar's max width constraint.
+    final double maxScale = size.width != 0.0
+      ? clampDouble((navBarConstraints.maxWidth - _kNavBarEdgePadding) / size.width, 1.0, 1.1)
+      : 1.1;
+
+    // This scale is estimated from the settings app in iOS 14.
+    // The large title scales linearly from 1.0 up to 1.15 magnification.
+    // The `navBarConstraints.maxHeight` value is the height of the nav bar,
+    // and `maxExtent` is the default large title height the nav bar snaps back to.
+    // The difference between the two heights is used to scale the title.
+    final double scale = clampDouble(1.0 + (navBarConstraints.maxHeight - maxExtent) / maxExtent * 0.12, 1.0, maxScale);
+
+    final Alignment resolvedAlignment = AlignmentDirectional.bottomStart.resolve(textDirection);
+
+    final Matrix4 resultMatrix = Matrix4.identity();
+
+    Offset? translation;
+    translation = resolvedAlignment.alongSize(size);
+
+    resultMatrix.translate(translation.dx, translation.dy);
+    resultMatrix.scale(scale, scale, 1.0);
+    resultMatrix.translate(-translation.dx, -translation.dy);
+
+    return resultMatrix;
+  }
+
+  @override
+  void performLayout() {
+    child?.layout(constraints, parentUsesSize: true);
+    size = child!.size;
+
+    _transform = _effectiveTransform;
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    context.pushTransform(
+      needsCompositing,
+      offset,
+      _transform!,
+      super.paint,
+      oldLayer: layer as TransformLayer?,
+    );
+  }
+
+  @override
+  void applyPaintTransform(RenderBox child, Matrix4 transform) {
+    transform.multiply(_transform!);
+  }
+
+  @override
+  bool hitTestChildren(BoxHitTestResult result, {required Offset position}) {
+    if (size.isEmpty || (child?.size.isEmpty ?? false)) {
```

Comment:
nit: you probably don't need these in `hitTestChildren` since `hitTest` has probably already done something similar